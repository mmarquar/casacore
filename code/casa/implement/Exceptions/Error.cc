//# Error.cc: Base class for all AIPS++ errors (templated classes)
//# Copyright (C) 1993,1994,1995,1996
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//#
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//#
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id$

#include <aips/Exceptions/Error.h>
#include <iostream.h>


//#  This template is for index errors  where the Index can be any type. It
//#  is derived from the class IndexError. This derivation allows the user
//#  to catch all index errors with one catch statement. The remainder of
//#  of the class is very similar to the AipsError class.
rtti_imp_init_a1(indexError);
rtti_imp_mbrf_a1(indexError);

template<class t>
indexError<t>::indexError(t oI, const Char *str) : oIndex(oI), IndexError(str)
{}
template<class t>
indexError<t>::indexError(t oI, const String &str) : oIndex(oI), IndexError(str)
{}

template<class t>
indexError<t>::indexError(ExcpError *excp) : IndexError(excp)
{
    indexError<t> *tmp;
    PCAST(tmp,indexError<t>,excp);
    if (tmp) {
        oIndex = (*tmp).oIndex;
        _equal = True;
    } else {
        _equal = False;
    }
}

template<class t>
indexError<t>::~indexError()
    {}

//#  This template is for duplicate key errors  where the Key can be any type. It
//#  is derived from the class DuplError. This derivation allows the user
//#  to catch all index errors with one catch statement. The remainder of
//#  of the class is very similar to the AipsError class.
rtti_imp_init_a1(duplError);
rtti_imp_mbrf_a1(duplError);

template<class t>
duplError<t>::duplError(ExcpError *excp) : DuplError(excp)
{
    duplError<t> *tmp;
    PCAST(tmp,duplError<t>,excp);
    if (tmp) {
        oKey = (*tmp).oKey;
        _equal = True;
    } else {
        _equal = False;
    }
}

template<class t>
duplError<t>::~duplError()
    {}


template<class t>
duplError<t>::duplError(t oI, const Char *str) : oKey(oI), DuplError(str)
{}
template<class t>
duplError<t>::duplError(t oI, const String &str) : oKey(oI), DuplError(str)
{}

