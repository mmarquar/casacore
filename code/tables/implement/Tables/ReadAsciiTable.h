//# AsciiFileIO.h: Filling a table from an Ascii file
//# Copyright (C) 1993,1994,1995
//# Associated Universities, Inc. Washington DC, USA.
//# 
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//# 
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//# 
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//# 
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id$

#if !defined (AIPS_ASCIIFILEIO_H)
#define AIPS_ASCIIFILEIO_H

#include <aips/aips.h>

#if defined(_AIX)
#pragma implementation ("AsciiFileIO.cc")
#endif

// <summary>
// Filling a table from an Ascii file.
// </summary>

// <use visibility=export>

// <prerequisites>
//  <li> <linkto class="Table:description">Table</linkto>
// <prerequisites>

// <synopsis>
// Global functions to fill a table from an Ascii file.
//
// The table columns are filled from a file containing the data values
// separated by whitespace (one line per table row).
// The following two lines define the columns in the table:
// <ol>
//   <li> The first line contains the names of the variables in each column.
//        These names may be enclosed in quotes (either single or double).     
//   <li> The second line contains the data types of each column.
//        Valid types are:
//      <ul>
//        <li>            R     for Real data
//        <li>            D     for Double Precision data
//        <li>            I     for Integer data
//        <li>            X     for Complex data (Real followed by Imaginary)
//        <li>            Z     for Complex data (Amplitude then Phase)
//        <li>            A     for ASCII data (must be enclosed in quotes
//                              if it contains one or more blanks)
//      </ul>
// </ol>
//
// There are two forms of the readAsciiTable function:
// <ol>
//  <li> The simplest form has one input file. The first two lines
//       define the columns as described above. The remaining lines contain
//       the data. For example:
//       <srcblock>
//        COLI   COLF   COLD       COLX        COLZ       COLS
//         I      R      D          X           Z          A
//        1      1.1    1.11       1.12 1.13   1.14 1.15  Str1
//        10     11     12         13   14     15   16    String17
//       </srcblock>
//       results in a table with 6 columns and 2 rows.
//  <li> The other form has two input files allowing the definition of
//       table keywords as well. The second input file contains the
//       column data. The first input file contains the keywords (if any)
//       and the column definitions.
//       The keywords in the first file, if there are any, must be enclosed
//       between a line that contains ".keywords" and a line that contains
//       ".endkeywords". Between these two lines each line should contain
//       the following:
//       <ul>
//        <li> The keyword name, e.g., ANYKEY
//        <li> The datatype of the keyword (cf. list of valid types above)
//        <li> The value or values for the keyword (the keyword may contain a
//             scalar or a vector of values).  e.g., 3.14159  21.78945
//      </ul>
//      For example:
//      <srcblock>
//       .keywords
//       KEYI  I  10
//       KEYIV I  11 12 13 14
//       KEYF  R  1.2
//       KEYFV R  -3.2 0 5.6
//       KEYD  D  1.23456789
//       KEYDV D  1 2 3 4 5 6 7 8 9
//       KEYX  X  -1.5 -3
//       KEYXC X  0 1 2 3 4 5 6 7 8 9
//       KEYZ  Z  -3  -1.5
//       KEYZV Z  0 0.1 0.2 0.3 0.4 0.5
//       KEYS  A  "1 2 3 4 5"
//       KEYSV A  " 1 2 " "AAA" BBB bbb CCc C "@#$%^&*()"
//       .endkeywords
//       COLI   COLF   COLD       COLX        COLZ       COLS
//        I      R      D          X           Z          A
//      </srcblock>
//      defines a table with 12 keywords (of which 6 contain vector values)
//      and 6 columns. The number of rows is determined by the number of
//      lines in the second input file.
// </ol>
// </synopsis>

// <example>
// <srcblock>
//   readAsciiTable ("file.in", "", "table.test");
// </srcblock>
// creates a table with name <src>table.test</src> from the text file
// <src>file.in</src>. The text file could look like:
// <srcblock>
//  COLI   COLF   COLD       COLX        COLZ       COLS
//   I      R      D          X           Z          A
//  1      1.1    1.11       1.12 1.13   1.14 1.15  Str1
//  10     11     12         13   14     15   16    String17
// </srcblock>
// resulting in a table with 6 columns and 2 rows.
// </example>

// <group name=readAsciiTable>


// Create a table with name as given by tableName. The layout of the
// table is defined in the first 2 lines of the input file. The remaining
// lines in the input file contain the data.
//
// When the tableDescName is not blank, the table description will
// be stored in a table description file with the given name.
void readAsciiTable (const Char* filein, const Char* tableDescName,
		     const Char* tableName);

// This form reads TWO Ascii files. The first file may contain 
// keywords and their values as well as the two lines described above for
// the names and type of variables.  The second file is intended for data only.
//
// When the tableDescName is not blank, the table description will
// be stored in a table description file with the given name.
void readAsciiTable (const Char* headerFile, const Char* dataFfile, 
		     const Char* tableDescName, const Char* tablename);

// </group>


#endif
