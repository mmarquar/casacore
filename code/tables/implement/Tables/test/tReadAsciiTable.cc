//# tReadAsciiTable.cc: Test program for the ReadAsciiTable functions
//# Copyright (C) 1994,1995,1996,1999,2000,2001
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This program is free software; you can redistribute it and/or modify it
//# under the terms of the GNU General Public License as published by the Free
//# Software Foundation; either version 2 of the License, or (at your option)
//# any later version.
//#
//# This program is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
//# more details.
//#
//# You should have received a copy of the GNU General Public License along
//# with this program; if not, write to the Free Software Foundation, Inc.,
//# 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id$

#include <aips/Tables/ReadAsciiTable.h>
#include <aips/Tables/TableDesc.h>
#include <aips/Tables/Table.h>
#include <aips/Tables/ScalarColumn.h>
#include <aips/Tables/TableRecord.h>
#include <aips/Arrays/Array.h>
#include <aips/Arrays/ArrayIO.h>
#include <aips/Exceptions/Error.h>
#include <aips/iostream.h>

// <summary> Test program for the ReadAsciiTable functions </summary>

// This program tests the functions in ReadAsciiTable.h.
// It uses some files in the test directory. The directory of those
// files is given in argv[1].
// The results are written to stdout. The script executing this program,
// compares the results with the reference output file.

void a (const String& dir);
void a1 (const String& dir);
void b (const String& dir, const String& postfix, Char separator);
void b1 (const String& dir);
void b2 (const String& dir);

int main (int argc, char** argv) {
    try {
	String dir;
	if (argc > 1) {
	    dir = argv[1];
	}
	a (dir);
	a1 (dir);
	b (dir, "", ' ');
	b (dir, "c", ',');
	b1 (dir);
	b2 (dir);
    } catch (AipsError x) {
	cout << "Caught an exception: " << x.getMesg() << endl;
	return 1;
    } 
    return 0;                           // exit with success status
}

void a (const String& dir)
{
    cout << ">>>" << endl;
    String formStr = readAsciiTable (dir + "tReadAsciiTable.in_tab", "",
				     "tReadAsciiTable_tmp.data_tab");
    cout << "<<<" << endl;
    cout << "Input format: [" << formStr << ']' << endl;
    Table tab("tReadAsciiTable_tmp.data_tab");
    cout << endl;
    cout << tab.nrow() << " rows, " << tab.tableDesc().ncolumn()
	 << " columns" << endl;
    ROScalarColumn<Int>     coli (tab,"COLI");
    ROScalarColumn<float>   colf (tab,"COLF");
    ROScalarColumn<double>  cold (tab,"COLD");
    ROScalarColumn<Complex> colx (tab,"COLX");
    ROScalarColumn<Complex> colz (tab,"COLZ");
    ROScalarColumn<String>  cols (tab,"COLS");
    for (uInt i=0; i<tab.nrow(); i++) {
	cout << coli(i) << " " << colf(i) << " " << cold(i) << " "
	     << colx(i) << " " << colz(i) << " " << cols(i) << endl;
    }
}

void a1 (const String& dir)
{
    cout << ">>>" << endl;
    String formStr = readAsciiTable (dir + "tReadAsciiTable.in_tah", "",
				     "tReadAsciiTable_tmp.data_tah", True);
    cout << "<<<" << endl;
    cout << "Input format: [" << formStr << ']' << endl;
    Table tab("tReadAsciiTable_tmp.data_tah");
    cout << endl;
    cout << tab.nrow() << " rows, " << tab.tableDesc().ncolumn()
	 << " columns" << endl;
    ROScalarColumn<Int>      col1 (tab,"Column1");
    ROScalarColumn<double>   col2 (tab,"Column2");
    ROScalarColumn<double>   col3 (tab,"Column3");
    ROScalarColumn<double>   col4 (tab,"Column4");
    ROScalarColumn<double>   col5 (tab,"Column5");
    ROScalarColumn<double>   col6 (tab,"Column6");
    ROScalarColumn<double>   col7 (tab,"Column7");
    ROScalarColumn<String>   col8 (tab,"Column8");
    for (uInt i=0; i<tab.nrow(); i++) {
	cout << col1(i) << " " << col2(i) << " " << col3(i) << " "
	     << col4(i) << " " << col5(i) << " " << col6(i) << " "
	     << col7(i) << " " << col8(i) << endl;
    }
}

void b (const String& dir, const String& postfix, Char separator)
{
    cout << ">>>" << endl;
    String formStr = readAsciiTable (dir + "tReadAsciiTable.in_tkh" + postfix,
				     dir + "tReadAsciiTable.in_tkd" + postfix,
				     "tReadAsciiTable_tmp",
				     "tReadAsciiTable_tmp.data_tk",
				     separator);
    cout << "<<<" << endl;
    cout << "Input format: [" << formStr << ']' << endl;
    cout << endl;
    TableDesc tabdesc("tReadAsciiTable_tmp");
    tabdesc.show();
    cout << endl;
    Table tab("tReadAsciiTable_tmp.data_tk");
    const TableRecord& keys = tab.keywordSet();
    cout << keys.description();
    cout << "KEYS " << keys.asShort ("KEYS") << endl;
    cout << "KEYI " << keys.asInt ("KEYI") << endl;
    cout << "KEYF " << keys.asfloat ("KEYF") << endl;
    cout << "KEYD " << keys.asdouble ("KEYD") << endl;
    cout << "KEYX " << keys.asComplex ("KEYX") << endl;
    cout << "KEYZ " << keys.asComplex ("KEYZ") << endl;
    cout << "KEYDX " << keys.asDComplex ("KEYDX") << endl;
    cout << "KEYDZ " << keys.asDComplex ("KEYDZ") << endl;
    cout << "KEYA " << keys.asString ("KEYA") << endl;
    cout << "KEYB " << keys.asBool ("KEYB") << endl;
    cout << "KEYSV " << keys.asArrayShort ("KEYSV") << endl;
    cout << "KEYIV " << keys.asArrayInt ("KEYIV") << endl;
    cout << "KEYFV " << keys.asArrayfloat ("KEYFV") << endl;
    cout << "KEYDV " << keys.asArraydouble ("KEYDV") << endl;
    cout << "KEYXC " << keys.asArrayComplex ("KEYXC") << endl;
    cout << "KEYZV " << keys.asArrayComplex ("KEYZV") << endl;
    cout << "KEYDXC " << keys.asArrayDComplex ("KEYDXC") << endl;
    cout << "KEYDZV " << keys.asArrayDComplex ("KEYDZV") << endl;
    cout << "KEYAV " << keys.asArrayString ("KEYAV") << endl;
    cout << "KEYBV " << keys.asArrayBool ("KEYBV") << endl;
    cout << endl;
    {
      ROTableColumn tabcol (tab, "COLI");
      const TableRecord& keycol = tabcol.keywordSet();
      cout << keycol.description();
      cout << "IKEYS " << keycol.asString ("IKEYS") << endl;
    }
    {
      ROTableColumn tabcol (tab, "COLDX");
      const TableRecord& keycol = tabcol.keywordSet();
      cout << keycol.description();
      cout << "IKEYS " << keycol.asString ("IKEYS") << endl;
      cout << "DKEYS " << keycol.asString ("DKEYS") << endl;
    }
    cout << tab.nrow() << " rows, " << tab.tableDesc().ncolumn()
	 << " columns" << endl;
    ROScalarColumn<Short>    cols (tab,"COLS");
    ROScalarColumn<Int>      coli (tab,"COLI");
    ROScalarColumn<float>    colf (tab,"COLF");
    ROScalarColumn<double>   cold (tab,"COLD");
    ROScalarColumn<Complex>  colx (tab,"COLX");
    ROScalarColumn<Complex>  colz (tab,"COLZ");
    ROScalarColumn<DComplex> coldx (tab,"COLDX");
    ROScalarColumn<DComplex> coldz (tab,"COLDZ");
    ROScalarColumn<String>   cola (tab,"COLA");
    ROScalarColumn<Bool>     colb (tab,"COLB");
    for (uInt i=0; i<tab.nrow(); i++) {
	cout << cols(i) << " " << coli(i) << " " << colf(i) << " "
	     << cold(i) << " " << colx(i) << " " << coldx(i) << " "
	     << colz(i) << " " << coldz(i) << " " << cola(i) << " "
	     << colb(i) << endl;
    }
}

void b1 (const String& dir)
{
    cout << ">>>" << endl;
    String formStr = readAsciiTable (dir + "tReadAsciiTable.in_tkh", "",
				     "tReadAsciiTable_tmp.data_tk");
    cout << "<<<" << endl;
    cout << "Input format: [" << formStr << ']' << endl;
    cout << endl;
    Table tab("tReadAsciiTable_tmp.data_tk");
    const TableRecord& keys = tab.keywordSet();
    cout << keys.description();
    cout << "KEYS " << keys.asShort ("KEYS") << endl;
    cout << "KEYI " << keys.asInt ("KEYI") << endl;
    cout << "KEYF " << keys.asfloat ("KEYF") << endl;
    cout << "KEYD " << keys.asdouble ("KEYD") << endl;
    cout << "KEYX " << keys.asComplex ("KEYX") << endl;
    cout << "KEYZ " << keys.asComplex ("KEYZ") << endl;
    cout << "KEYDX " << keys.asDComplex ("KEYDX") << endl;
    cout << "KEYDZ " << keys.asDComplex ("KEYDZ") << endl;
    cout << "KEYA " << keys.asString ("KEYA") << endl;
    cout << "KEYB " << keys.asBool ("KEYB") << endl;
    cout << "KEYSV " << keys.asArrayShort ("KEYSV") << endl;
    cout << "KEYIV " << keys.asArrayInt ("KEYIV") << endl;
    cout << "KEYFV " << keys.asArrayfloat ("KEYFV") << endl;
    cout << "KEYDV " << keys.asArraydouble ("KEYDV") << endl;
    cout << "KEYXC " << keys.asArrayComplex ("KEYXC") << endl;
    cout << "KEYZV " << keys.asArrayComplex ("KEYZV") << endl;
    cout << "KEYDXC " << keys.asArrayDComplex ("KEYDXC") << endl;
    cout << "KEYDZV " << keys.asArrayDComplex ("KEYDZV") << endl;
    cout << "KEYAV " << keys.asArrayString ("KEYAV") << endl;
    cout << "KEYBV " << keys.asArrayBool ("KEYBV") << endl;
    cout << endl;
    {
      ROTableColumn tabcol (tab, "COLI");
      const TableRecord& keycol = tabcol.keywordSet();
      cout << keycol.description();
      cout << "IKEYS " << keycol.asString ("IKEYS") << endl;
    }
    {
      ROTableColumn tabcol (tab, "COLDX");
      const TableRecord& keycol = tabcol.keywordSet();
      cout << keycol.description();
      cout << "IKEYS " << keycol.asString ("IKEYS") << endl;
      cout << "DKEYS " << keycol.asString ("DKEYS") << endl;
    }
    cout << tab.nrow() << " rows, " << tab.tableDesc().ncolumn()
	 << " columns" << endl;
    ROScalarColumn<Short>    cols (tab,"COLS");
    ROScalarColumn<Int>      coli (tab,"COLI");
    ROScalarColumn<float>    colf (tab,"COLF");
    ROScalarColumn<double>   cold (tab,"COLD");
    ROScalarColumn<Complex>  colx (tab,"COLX");
    ROScalarColumn<Complex>  colz (tab,"COLZ");
    ROScalarColumn<DComplex> coldx (tab,"COLDX");
    ROScalarColumn<DComplex> coldz (tab,"COLDZ");
    ROScalarColumn<String>   cola (tab,"COLA");
    ROScalarColumn<Bool>     colb (tab,"COLB");
    for (uInt i=0; i<tab.nrow(); i++) {
	cout << cols(i) << " " << coli(i) << " " << colf(i) << " "
	     << cold(i) << " " << colx(i) << " " << colz(i) << " "
	     << coldx(i) << " " << coldz(i) << " " << cola(i) << " "
	     << colb(i) << endl;
    }
}

void b2 (const String& dir)
{
    cout << ">>>" << endl;
    String formStr = readAsciiTable (dir + "tReadAsciiTable.in_tkh", "",
				     "tReadAsciiTable_tmp.data_tk", True);
    cout << "<<<" << endl;
    cout << "Input format: [" << formStr << ']' << endl;
    cout << endl;
    Table tab("tReadAsciiTable_tmp.data_tk");
    const TableRecord& keys = tab.keywordSet();
    cout << keys.description();
    cout << "KEYS " << keys.asShort ("KEYS") << endl;
    cout << "KEYI " << keys.asInt ("KEYI") << endl;
    cout << "KEYF " << keys.asfloat ("KEYF") << endl;
    cout << "KEYD " << keys.asdouble ("KEYD") << endl;
    cout << "KEYX " << keys.asComplex ("KEYX") << endl;
    cout << "KEYZ " << keys.asComplex ("KEYZ") << endl;
    cout << "KEYDX " << keys.asDComplex ("KEYDX") << endl;
    cout << "KEYDZ " << keys.asDComplex ("KEYDZ") << endl;
    cout << "KEYA " << keys.asString ("KEYA") << endl;
    cout << "KEYB " << keys.asBool ("KEYB") << endl;
    cout << "KEYSV " << keys.asArrayShort ("KEYSV") << endl;
    cout << "KEYIV " << keys.asArrayInt ("KEYIV") << endl;
    cout << "KEYFV " << keys.asArrayfloat ("KEYFV") << endl;
    cout << "KEYDV " << keys.asArraydouble ("KEYDV") << endl;
    cout << "KEYXC " << keys.asArrayComplex ("KEYXC") << endl;
    cout << "KEYZV " << keys.asArrayComplex ("KEYZV") << endl;
    cout << "KEYDXC " << keys.asArrayDComplex ("KEYDXC") << endl;
    cout << "KEYDZV " << keys.asArrayDComplex ("KEYDZV") << endl;
    cout << "KEYAV " << keys.asArrayString ("KEYAV") << endl;
    cout << "KEYBV " << keys.asArrayBool ("KEYBV") << endl;
    cout << endl;
    cout << tab.nrow() << " rows, " << tab.tableDesc().ncolumn()
	 << " columns" << endl;
    ROScalarColumn<String> col1 (tab,"Column1");
    ROScalarColumn<String> col2 (tab,"Column2");
    ROScalarColumn<String> col3 (tab,"Column3");
    ROScalarColumn<String> col4 (tab,"Column4");
    ROScalarColumn<String> col5 (tab,"Column5");
    ROScalarColumn<String> col6 (tab,"Column6");
    ROScalarColumn<String> col7 (tab,"Column7");
    ROScalarColumn<String> col8 (tab,"Column8");
    ROScalarColumn<String> col9 (tab,"Column9");
    ROScalarColumn<String> col10 (tab,"Column10");
    for (uInt i=0; i<tab.nrow(); i++) {
	cout << col1(i) << " " << col2(i) << " " << col3(i) << " "
	     << col4(i) << " " << col5(i) << " " << col6(i) << " "
	     << col7(i) << " " << col8(i) << " " << col9(i) << " "
	     << col10(i) << endl;
    }
}
