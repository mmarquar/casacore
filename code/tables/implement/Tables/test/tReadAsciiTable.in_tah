1      1.1    1.11       1.12 1.13   1.14 1.15  ""
10     11     12         13   14     15   16    str2
