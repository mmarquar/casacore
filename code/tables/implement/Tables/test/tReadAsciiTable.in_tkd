1      1.1    1.11       1.12 1.13   1.14 1.15  Str1
-1     -1.1   -1.11      -1.2 -1.3  -1.4 -1.5   "Str1 3"
10     11     12         13   14     15   16    String17
