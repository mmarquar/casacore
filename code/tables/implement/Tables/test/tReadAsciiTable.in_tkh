.keywords
KEYI  i  10
KEYIV I  11 12 13 14
KEYF  R  1.2
KEYFV r  -3.2 0 5.6
KEYD  d  1.23456789
KEYDV D  1 2 3 4 5 6 7 8 9
KEYX  X  -1.5 -3
KEYXC x  0 1 2 3 4 5 6 7 8 9
KEYZ  Z  -3  -1.5
KEYZV z  0 0.1 0.2 0.3 0.4 0.5
KEYS  a  "1 2 3 4 5"
KEYSV A  " 1 2 " "AAA" BBB bbb CCc C "@#$%^&*()"
.endkeywords
COLI   COLF   COLD       COLX        COLZ       COLS
 i      R      d          X           z          A
