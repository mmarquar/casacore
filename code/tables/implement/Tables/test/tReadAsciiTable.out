>>>
<<<
Input format: [COLI=I, COLF=R, COLD=D, COLX=X, COLZ=Z, COLS=A]

2 rows, 6 columns
1 1.1 1.11 (1.12, 1.13) (1.13977, 0.0228797) Str1
10 11 12 (13, 14) (14.4189, 4.13456) String17
>>>
<<<
Input format: [Column1=I, Column2=D, Column3=D, Column4=D, Column5=D, Column6=D, Column7=D, Column8=A]

2 rows, 8 columns
1 1.1 1.11 1.12 1.13 1.14 1.15 
10 11 12 13 14 15 16 str2
>>>
<<<
Input format: [COLS=S, COLI=I, COLF=R, COLD=D, COLX=X, COLDX=DX, COLZ=Z, COLDZ=DZ, COLA=A, COLB=B]


TableDesc tReadAsciiTable_tmp   version    (Directory ./)
---------
  Comment: 
  #Keywords = 0
  #Columns  = 10
   Name=COLS   DataType=Short
   DataManager=StManAipsIO/StManAipsIO   Default=0
   Comment = 
   #keywords=0

   Name=COLI   DataType=Int
   DataManager=StManAipsIO/StManAipsIO   Default=0
   Comment = 
   #keywords=0

   Name=COLF   DataType=float
   DataManager=StManAipsIO/StManAipsIO   Default=0
   Comment = 
   #keywords=0

   Name=COLD   DataType=double
   DataManager=StManAipsIO/StManAipsIO   Default=0
   Comment = 
   #keywords=0

   Name=COLX   DataType=Complex
   DataManager=StManAipsIO/StManAipsIO   Default=(0, 0)
   Comment = 
   #keywords=0

   Name=COLDX   DataType=DComplex
   DataManager=StManAipsIO/StManAipsIO   Default=(0, 0)
   Comment = 
   #keywords=0

   Name=COLZ   DataType=Complex
   DataManager=StManAipsIO/StManAipsIO   Default=(0, 0)
   Comment = 
   #keywords=0

   Name=COLDZ   DataType=DComplex
   DataManager=StManAipsIO/StManAipsIO   Default=(0, 0)
   Comment = 
   #keywords=0

   Name=COLA   DataType=String
   DataManager=StManAipsIO/StManAipsIO   Default=
   Comment = 
   #keywords=0

   Name=COLB   DataType=Bool
   DataManager=StManAipsIO/StManAipsIO   Default=0
   Comment = 
   #keywords=0


0  KEYS : Short
1  KEYSV : Array<Short>[-1]
2  KEYI : Int
3  KEYIV : Array<Int>[-1]
4  KEYF : float
5  KEYFV : Array<float>[-1]
6  KEYD : double
7  KEYDV : Array<double>[-1]
8  KEYX : Complex
9  KEYXC : Array<Complex>[-1]
10  KEYZ : Complex
11  KEYZV : Array<Complex>[-1]
12  KEYDX : DComplex
13  KEYDXC : Array<DComplex>[-1]
14  KEYDZ : DComplex
15  KEYDZV : Array<DComplex>[-1]
16  KEYA : String
17  KEYAV : Array<String>[-1]
18  KEYB : Bool
19  KEYBV : Array<Bool>[-1]
KEYS 20
KEYI 10
KEYF 1.2
KEYD 1.23457
KEYX (-1.5, -3)
KEYZ (-2.99897, 0.0785308)
KEYDX (-2.5, -4)
KEYDZ (-3.99619, 0.174478)
KEYA 1 2 3 4 5
KEYB 1
KEYSV [10, 11, 12]
KEYIV [11, 12, 13, 14]
KEYFV [-3.2, 0, 5.6]
KEYDV [1, 2, 3, 4, 5, 6, 7, 8, 9]
KEYXC [(0, 1), (2, 3), (4, 5), (6, 7), (8, 9)]
KEYZV [(0, 0), (0.199997, 0.00104719), (0.399985, 0.00349061)]
KEYDXC [(1, 2), (3, 4), (5, 6), (7, 8), (9, 10)]
KEYDZV [(0.0999994, 0.000349065), (0.299993, 0.00209438), (0.499973, 0.00523589)]
KEYAV [ 1 2 , AAA, BBB, bbb, CCc, C, @#$%^&*()]
KEYBV [0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1]

3 rows, 10 columns
0 1 1.1 1.11 (1.12, 1.13) (1.13977, 0.0228797) (2.12, 2.13) (2.13849, 0.0802838) Str1 1
-2 -1 -1.1 -1.11 (-1.2, -1.3) (-1.39952, 0.0366477) (-2.2, -2.3) (-2.39772, 0.104687) Str1 3 1
123 10 11 12 (13, 14) (14.4189, 4.13456) (23, -24) (22.4699, -10.9593)  0
